<!DOCTYPE html>
<html>
  <head>
    <title>Title</title>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"/>
    <style type="text/css">
      /* Slideshow styles */
      body {
        font-family: 'Helvetica';
      }
    </style>
  </head>
  <body>
    <textarea id="source">

class: center, middle

# Sci-Kit Learn for Biologists

Austin G. Davis-Richardson

Triplett Group, Microbiology &nbsp; Cell Science, University of Florida

---

# What is Sci-Kit Learn?

---

# SciKit-Learn

- Machine learning library for Python
  - Classification
  - Regression
  - Clustering
  - Preprocessing
  - Dimension Reduction
  - Cross-Validation

---

# Some Examples
  (with a little BioPython)

---

# Sequence Binner

- K Means Clustering
- Metagenome Binner

```python
from Bio import SeqIO
from sklearn.preprocessing import HashingVectorizer
from sklearn.clustering import KMeansClustering
from sklearn.pipeline import Pipeline

classifier = Pipeline(['hash', HashingVectorizer()],
                      ['classifier', KMeansClustering()])

# train classifier
with open(sys.argv[1]) as handle:
    records = SeqIO.parse(handle, 'fasta')
    features, labels = [ (x.id, str(x.seq)) for x in records ]

    classifier.fit(labels, features)

# classify new sequences
with open(sys.argv[2]) as handle:
    records = SeqIO.parse(handle, 'fasta')
    features, labels = [ (x.id, str(x.seq)) for x in records ]

    classifier.predict(labels, features)
```

---

# Metagenome Binner

---

# Outlier Removal
(One Class SVM)

---

# Artifact Detection

---

# Feature Selection

(which k-mers best separate genome A from genome B?)

---


    </textarea>
    <script src="http://gnab.github.io/remark/downloads/remark-0.6.0.min.js" type="text/javascript">
    </script>
    <script type="text/javascript">
      var slideshow = remark.create();
    </script>
  </body>
</html>
