<!DOCTYPE html>
<html>
  <head>
    <title>SciKit-Learn for Biologists</title>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"/>
    <style type="text/css">
      @import url(http://fonts.googleapis.com/css?family=Droid+Serif);
      @import url(http://fonts.googleapis.com/css?family=Yanone+Kaffeesatz);
      /* Slideshow styles */
      body {
        font-family: 'Droid Serif';
        font-size: 20px;
      }
      h1, h2, h3  {
        font-family: 'Yanone Kaffeesatz';
      }
      code {
        font-family: 'Menlo';
      }
      a {
        color: #00aaFF;
      }
      .footnote {
        position: absolute;
        bottom: 3em;
        font-size: 18px;
      }
      .pull-right {
        width: 75%;
        float: right;
        padding-top: 2em;
      }
      .inverse {
        background: #272822;
        color: #aaaaaa;
        text-shadow: 0 0 20px #333;
      }
      .inverse a {
        color: #FFaa00;
        text-decoration: none;
      }
      .inverse h1, .inverse h2, .inverse h3 {
        color: #efefef;
      }
      img {
        width: 100%;
        height: 100%;
      }
    </style>
  </head>
  <body>
    <textarea id="source">

class: center, middle, inverse

# Sci-Kit Learn for Biology

Austin G. Davis-Richardson

Triplett Group

Microbiology + Cell Science

University of Florida

.footnote[
slides: https://github.com/audy/presentations
]

---

class: center, middle, inverse

# What is Sci-Kit Learn?

---

# SciKit-Learn

![](media/ml_map.png)

.footnote[
http://peekaboo-vision.blogspot.com/2013/01/machine-learning-cheat-sheet-for-scikit.html
]

---

# SciKit-Learn

- **Supervised Learning**
    - Linear Models
    - Random Forests, Decision Trees, AdaBoost
    - Support Vector Machines
    - Outlier Detection (1-Class SVM)
- **Unsupervised Learning**
    - Clustering: Spectral, KMeans, MinBatch KMeans
    - Dimension Reduction (PCA, NMDS)
- **Evaluation**
    - Accuracy
    - ROC/AUC
    - RMSE
    - Silhouette Index
    - Cross-Validation
      - Nested/Stratified/Leave-K-Out/Leave-One-Label-Out
    - "Dummy" Algorithms
- **Parameter Tuning**
    - Grid Search
    - Randomized Search

*and much more!*
---

class: center, middle, inverse

# Advantages

---
# Consistent API

## Rapid Prototyping and Evaluation

- Algorithms expose a common API:
    1. `estimators` - regression
    2. `predictors` - classification/clustering
    3. `transformers` - dimension reduction, normalization, scaling
- Allows for **Rapid Prototyping**
    - Can easily swap an algorithm by changing the class.
- Easy to chain algorithms
    - transform => transform => predict

```python
classifiers = [
  DecisionTreeClassifier(),
  RandomForestClassifier(),
  SVC()
]

for classifier in classifiers:
  print cross_val_score(classifier, features, labels).mean()
```
---
# Consistent API

## Chaining Steps

```python
pipeline = Pipeline([('scale', StandardScaler()),
                     ('f_selection', SelectKBest()),
                     ('pca', PCA()),
                     ('classify', RandomForestClassifier())
                    ])

cross_val_score(pipeline, features, labels)
```

```python
pipeline = Pipeline([('scale', StandardScaler()),
                     ('f_selection', SelectKBest()),
                     ('mds', MDS()),
                     ('classify', LinearRegression())
                    ])

cross_val_score(pipeline, features, labels)
```

---

class: center, middle, inverse

# Well-Documented

---

![well documented](./media/sklearn-docu.png)

.footnote[http://scikit-learn.org/stable/auto_examples/plot_multioutput_face_completion.html]

---
class: center, middle, inverse

# Active

---

# Active

- Development occuring on GitHub
    - https://github.com/scikit-learn/scikit-learn/
- Still an early release: 0.14b
- Open Source: MITv3

![pulse](./media/pulse.png)

---

class: middle, center, inverse

# Some Examples

---

class: middle, center, inverse
# Example 1/4
## Separate Genome from Metagenome

---

class: middle, center, inverse
# Example 2/4
## 16S rRNA Sequence Classifier

---

class: middle, center, inverse
# Example 3/4
## Protein-Drug Binding Prediction

---

class: middle, center, inverse
# Example 4/4
## Enterotypes of the Human Gut

---
class: middle

# Additional Resources

- SciKit-Learn (sckit-learn.org)
- Pandas (pandas.pydata.org) - Data manipulation
- Numpy (numpy.org) - Fast, numerical calculations
- IPython (ipython.org) - Interactive Computing
- BioPython (biopython.org) - Integrate Bioinformatics tools

# More examples

- SciKit Learn w/ HapMap data
  http://nbviewer.ipython.org/github/glouppe/talk-sklearn-mloss-nips2013/blob/master/oral/sklearn-mloss.ipynb

.footnote[
slides: https://github.com/audy/presentations
]

    </textarea>
    <script src="http://gnab.github.io/remark/downloads/remark-0.6.0.min.js" type="text/javascript">
    </script>
    <script type="text/javascript">
      var slideshow = remark.create();
    </script>
  </body>
</html>
